### PR TITLE
RBAC- and proxy servlets

### PR DESCRIPTION
**Pull request by johanfylling**
Adding support for proxying RBAC-management and batch-queries via Jakarta servlets.

Implements STY-14741

### CODE REVIEW

**Comment by wusticality** [comment]
Particularly since these repositories are public-facing, I think we should write detailed descriptions for new PRs, I think a high-level overview of what the PR does would be helpful, not just for others but for me haha

File: core/src/main/java/com/styra/run/ApiClient.java [hunk 1]
```diff
@@ -1,5 +1,7 @@
 package com.styra.run;
 
+import com.styra.run.exceptions.RetryException;
```
Comment (wusticality) [inline core/src/main/java/com/styra/run/ApiClient.java:3]: Dead code? Didn't see any other changes in this file

File: core/src/main/java/com/styra/run/BlockingApiClient.java [hunk 2]
```diff
@@ -1,5 +1,6 @@
 package com.styra.run;
 
+import com.styra.run.exceptions.RetryException;
```
Comment (wusticality) [inline core/src/main/java/com/styra/run/BlockingApiClient.java:3]: Dead code?

File: sdk/src/main/java/com/styra/run/DefaultApiClient.java [hunk 3]
```diff
@@ -1,5 +1,8 @@
 package com.styra.run;
 
+import com.styra.run.exceptions.RetryException;
```
Comment (wusticality) [inline sdk/src/main/java/com/styra/run/DefaultApiClient.java:3]: Dead code?

File: core/src/main/java/com/styra/run/ApiClient.java (hunk 1 above)
Comment (johanfylling) [reply core/src/main/java/com/styra/run/ApiClient.java:3]: No, I've just moved this exception into a different package. Previously, it resided in the same package as this class, which doesn't require an import.

File: core/src/main/java/com/styra/run/BlockingApiClient.java (hunk 2 above)
Comment (johanfylling) [reply core/src/main/java/com/styra/run/BlockingApiClient.java:3]: No; same reason as above.

File: sdk/src/main/java/com/styra/run/DefaultApiClient.java (hunk 3 above)
Comment (johanfylling) [reply sdk/src/main/java/com/styra/run/DefaultApiClient.java:3]: No.

File: core/src/main/java/com/styra/run/session/TenantInputTransformer.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+        if (input != null && !input.isEmpty()) {
+            Map<?, ?> inputValue = Types.safeCast(Map.class, input.getValue());
+            if (inputValue == null) {
+                // The given input isn't a map, so we can't inject session info into it.
+                // Use existing input, and leave it up to the policy to reject or decline.
+                return input;
+            }
+            // Don't let client-side override or define session attributes
+            inputValue.entrySet().stream()
```
Comment (wusticality) [inline core/src/main/java/com/styra/run/session/TenantInputTransformer.java:22]: In the `golang` SDK I only inject the `tenant` and `session` keys if _both_ keys are unset

File: core/src/main/java/com/styra/run/session/TenantSession.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    private static final String SUBJECT_KEY = "subject";
+    private static final String TENANT_KEY = "tenant";
+
+    public TenantSession(String subject, String tenant) {
+        super(createMap(subject,
+                requireNotEmpty(tenant, "tenant must not be null or empty")));
+    }
+
+    private TenantSession(Map<String, Object> map) {
+        super(require(map, (m) -> m.containsKey(TENANT_KEY),
```
Comment (wusticality) [inline core/src/main/java/com/styra/run/session/TenantSession.java:28]: Shouldn't we require both keys here?

File: servlet/src/main/java/com/styra/run/session/CookieSessionManager.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Override
+    public TenantSession getSession(HttpServletRequest request) {
+        Cookie[] cookies = request.getCookies();
+        if (cookies != null) {
+            return Arrays.stream(cookies)
+                    .filter(cookie -> cookieName.equals(cookie.getName()))
+                    .findAny()
+                    .map(cookie -> {
+                        String[] components = cookie.getValue().split("\\s*/\\s*", 2);
```
Comment (wusticality) [inline servlet/src/main/java/com/styra/run/session/CookieSessionManager.java:25]: Should probably only look for the cookie named `"user"` here, no?

File: core/src/main/java/com/styra/run/Proxy.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+ * <p>
+ * By default, non-boolean policy decisions are rejected, and replaced with an error;
+ * this behaviour can be toggled off.
+ * <p>
+ * By default, result attributes are dropped.
+ *
+ * @param <S> the {@link Session session} type
+ */
+// TODO: Configurable allow-list of rule paths.
+public class Proxy<S extends Session> {
```
Comment (wusticality) [inline core/src/main/java/com/styra/run/Proxy.java:31]: I'm not sure how `java` modules work but you may want to rename this file like `BatchQueryProxy` as you'll need to add proxies for the CRUD operations and also `Query` / `Check`. I'd probably put them in a subfolder at minimum

File: core/src/main/java/com/styra/run/rbac/RbacManager.java [hunk 8]
```diff
@@ -1,87 +1,165 @@
...
+ * Manager of RBAC bindings maintained in Styra Run.
+ */
 public class RbacManager {
     public static final String AUTHZ_PATH = "rbac/manage/allow";
     private static final String ROLES_PATH = "rbac/roles";
     private static final String USER_BINDINGS_PATH = "rbac/user_bindings";
-    private static final String USER_BINDINGS_FORMAT = USER_BINDINGS_PATH + "/%s";
-    private static final String USER_BINDING_FORMAT = USER_BINDINGS_FORMAT + "/%s";
 
     private final StyraRun styraRun;
```
Comment (wusticality) [inline core/src/main/java/com/styra/run/rbac/RbacManager.java:28]: To be consistent with the `golang` SDK, I wonder if we shouldn't rename this type to `client` or something similar

File: core/src/main/java/com/styra/run/rbac/RbacManager.java [hunk 9]
```diff
@@ -1,87 +1,165 @@
...
+    }
+
+    /**
+     * Set the {@link UserBinding user-binding} for a {@link User}.
+     *
+     * @param userBinding the {@link UserBinding} to set
+     * @param session the {@link TenantSession session} information authorizing the request for a given tenant
+     * @return a {@link CompletableFuture} for the action
+     */
+    public CompletableFuture<Void> setUserBinding(UserBinding userBinding, TenantSession session) {
```
Comment (wusticality) [inline core/src/main/java/com/styra/run/rbac/RbacManager.java:67]: To be consistent with the `golang` SDK and our docs, I'd recommend calling this `putUserBinding`

File: core/src/main/java/com/styra/run/rbac/RbacManager.java [hunk 10]
```diff
@@ -1,87 +1,165 @@
...
+
+    /**
+     * Lists all {@link UserBinding user-bindings} for the tenant in the provided {@link TenantSession session}.
+     * <p>
+     * The returned list of user-bindings is sorted lexicographically by {@link User#getId() id}.
+     *
+     * @param session the {@link TenantSession session} information authorizing the request for a given tenant
+     * @return a {@link CompletableFuture} that completes to a list of all {@link UserBinding user-bindings} for the tenant
+     */
+    public CompletableFuture<List<UserBinding>> listUserBindings(TenantSession session) {
```
Comment (wusticality) [inline core/src/main/java/com/styra/run/rbac/RbacManager.java:100]: I'd call these `listUserBindingsAll` and `listUserBindings` to be consistent

File: core/src/test/groovy/com/styra/run/rbac/RbacManagerSpec.groovy [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+import static com.styra.run.ApiClient.Method.PUT
+import static com.styra.run.test.apiClients.mockApiClient
+import static com.styra.run.test.apiClients.reply
+import static com.styra.run.test.apiClients.response
+
+class RbacManagerSpec extends Specification {
+    @Unroll
+    def "RBAC-Manager can fetch roles"() {
+        def baseUri = 'https://example.com/my/proj/env'
+        def authzInput = new TenantSession('alice', 'AcmeCorp')
```
Comment (wusticality) [inline core/src/test/groovy/com/styra/run/rbac/RbacManagerSpec.groovy:22]: Should `AcmeCorp` be lowercase?

**Comment by wusticality** [comment]
It won't let me comment on parts of the `README` for some reason, so dropping a few comments here:

* Seems like it should be possible to boot the client without environment variables, no?
* I'd call the client `Proxy` section `BatchQuery`.
* You may want to reorder the various sections to mirror the `golang` `README`.
* I think you should document the RBAC functions as well. The `golang` `README` also shows example outputs so the user can get a feel for how the request payload / response looks.

**Review by wusticality — CHANGES_REQUESTED** [review]
Nice work! Some questions / comments / changes

File: core/src/main/java/com/styra/run/session/TenantInputTransformer.java (hunk 4 above)
Comment (johanfylling) [reply core/src/main/java/com/styra/run/session/TenantInputTransformer.java:22]: When this tenant/subject strategy is in effect, I think it's a mistake to let the client-side dictate relevant session info under _any_ circumstances. An attacker could simply opt to not send the session cookie, and just inject whatever session info they want to spoof into all the query inputs.

File: servlet/src/main/java/com/styra/run/session/CookieSessionManager.java (hunk 6 above)
Comment (johanfylling) [reply servlet/src/main/java/com/styra/run/session/CookieSessionManager.java:25]: That's done above, on line 22. The cookie name is configurable.

File: core/src/main/java/com/styra/run/Proxy.java (hunk 7 above)
Comment (johanfylling) [reply core/src/main/java/com/styra/run/Proxy.java:31]: Yea. I'm making some of those changes in my task to integrate the testing-framework. I'll include these modifications in that PR.

File: core/src/main/java/com/styra/run/rbac/RbacManager.java (hunk 8 above)
Comment (johanfylling) [reply core/src/main/java/com/styra/run/rbac/RbacManager.java:28]: I feel `client` is a bit too generic. In Node, we also call it `StyraRun`. Perhaps `StyraRunClient` is a good middle-ground.
The way Java deals with imports, makes too generic names that may overlap with other libraries a huge pain.

File: core/src/main/java/com/styra/run/rbac/RbacManager.java (hunk 9 above)
Comment (johanfylling) [reply core/src/main/java/com/styra/run/rbac/RbacManager.java:67]: Sure, I'll make that change.

What are "our docs", btw? We have that google-doc with the SDK overview. We discussed transcribing that into markdown and putting it into the test-framework repo; but I haven't seen that.

File: core/src/main/java/com/styra/run/rbac/RbacManager.java (hunk 10 above)
Comment (johanfylling) [reply core/src/main/java/com/styra/run/rbac/RbacManager.java:100]: Java lends itself to function-overloading, but I can rename this to `listAllUserBindings`, to be consistent (doesn't that have a better flow to it?).

I'm thinking:

* `listAllUserBindings()`: for fetching all bindings known to Styra Run. To be avoided in systems with many bindings. This probably needs to be changed to eventually return a cursor of some sort, that pumps pages as-defined by Styra Run.
* `listUserBindings()`: for fetching paginated bindings known to Styra Run. User can define pagination and search parameters; both of which needs support in Styra Run.
* `getUserBinsings()`:  takes a list of users, for which we'll fetch bindings.

I don't recall exactly how our previous discussion regarding this went, to be honest 😅 .

**Comment by johanfylling** [comment]
I intentionally didn't put too much effort into the README, as I think we should go through all of the SDK READMEs to define an appropriate template to follow everywhere.

File: core/src/main/java/com/styra/run/session/TenantInputTransformer.java (hunk 4 above)
Comment (wusticality) [reply core/src/main/java/com/styra/run/session/TenantInputTransformer.java:22]: Ah, I see. I'll fix that in the `golang` SDK as well

**Review by wusticality — APPROVED** [review]
lgtm

**Pull request merged**